src/ReadonlyArray.ts

```diff
@@ -0,0 +1,1678 @@
+/**
+ * @since 2.5.0
+ */
+import { Alternative1 } from './Alternative'
+import { Applicative } from './Applicative'
+import { Compactable1, Separated } from './Compactable'
+import { Either } from './Either'
+import { Eq } from './Eq'
+import { Extend1 } from './Extend'
+import { FilterableWithIndex1 } from './FilterableWithIndex'
+import { Foldable1 } from './Foldable'
+import { FoldableWithIndex1 } from './FoldableWithIndex'
+import { Predicate, Refinement, identity } from './function'
+import { FunctorWithIndex1 } from './FunctorWithIndex'
+import { HKT } from './HKT'
+import { Monad1 } from './Monad'
+import { Monoid } from './Monoid'
+import { ReadonlyNonEmptyArray } from './ReadonlyNonEmptyArray'
+import { isSome, none, Option, some } from './Option'
+import { fromCompare, getMonoid as getOrdMonoid, Ord, ordNumber } from './Ord'
+import { pipeable } from './pipeable'
+import { Show } from './Show'
+import { TraversableWithIndex1 } from './TraversableWithIndex'
+import { Unfoldable1 } from './Unfoldable'
+import { Witherable1 } from './Witherable'
+
+declare module './HKT' {
+  interface URItoKind<A> {
+    readonly ReadonlyArray: ReadonlyArray<A>
+  }
+}
+
+/**
+ * @since 2.5.0
+ */
+export const URI = 'ReadonlyArray'
+
+/**
+ * @since 2.5.0
+ */
+export type URI = typeof URI
+
+/**
+ * @since 2.5.0
+ */
+// tslint:disable-next-line: readonly-array
+export function fromArray<A>(as: Array<A>): ReadonlyArray<A> {
+  const l = as.length
+  if (l === 0) {
+    return empty
+  }
+  const ras = Array(l)
+  for (let i = 0; i < l; i++) {
+    ras[i] = as[i]
+  }
+  return ras
+}
+
+/**
+ * @since 2.5.0
+ */
+// tslint:disable-next-line: readonly-array
+export function toArray<A>(ras: ReadonlyArray<A>): Array<A> {
```

Comment:
faster than `slice`?

Reply:
based on https://github.com/gcanti/fp-ts/blob/5684ae1919044267d2b5a706800fd528bf6afc51/perf/Array/copy.ts seems a bit faster

Reply:
ok

Reply:
This actually made me frowning, so I played with that perf. Here the results on my machine (Linux) with Node.js 12.8:

Same perf from the repo, meaning array with 3 elements:
```
copy (native) x 42,886,253 ops/sec ±2.95% (80 runs sampled)
copy (Array) x 35,017,908 ops/sec ±2.70% (80 runs sampled)
Fastest is copy (native)
```

Array with 20 elements:
```
copy (native) x 22,314,806 ops/sec ±1.86% (90 runs sampled)
copy (Array) x 16,915,433 ops/sec ±1.20% (86 runs sampled)
Fastest is copy (native)
```

Array with 100 elements:
```
copy (native) x 8,201,057 ops/sec ±0.53% (92 runs sampled)
copy (Array) x 4,548,024 ops/sec ±0.93% (91 runs sampled)
Fastest is copy (native)
```

Array with 1k elements:
```
copy (native) x 843,804 ops/sec ±0.53% (94 runs sampled)
copy (Array) x 438,358 ops/sec ±1.06% (92 runs sampled)
Fastest is copy (native)
```

I actually had some weird results from the case with the array with 3 elements, like this one:
```
copy (native) x 29,569,974 ops/sec ±2.14% (82 runs sampled)
copy (Array) x 36,031,090 ops/sec ±3.06% (86 runs sampled)
Fastest is copy (Array) 
```

But generally it seems that `.slice` tends to be as twice as faster then `copy`.

Reply:
@DenisFrezzato here's mine (OSX 10.9.6, **Node.js 10.15.3**) and using `range` to build the arrays

`range(1, 3)`

```
copy (native) x 32,922,319 ops/sec ±0.52% (90 runs sampled)
copy (Array) x 79,747,647 ops/sec ±0.60% (85 runs sampled)
Fastest is copy (Array)
```

`range(1, 20)`

```
copy (native) x 17,306,844 ops/sec ±1.28% (89 runs sampled)
copy (Array) x 22,161,454 ops/sec ±1.71% (81 runs sampled)
Fastest is copy (Array)
```

`range(1, 100)`

```
copy (native) x 6,891,346 ops/sec ±1.01% (89 runs sampled)
copy (Array) x 5,313,439 ops/sec ±1.05% (87 runs sampled)
Fastest is copy (native)
```

`range(1, 1000)`

```
copy (native) x 893,195 ops/sec ±0.88% (87 runs sampled)
copy (Array) x 623,234 ops/sec ±1.01% (82 runs sampled)
Fastest is copy (native)
```

Reply:
apologies for asking this... but how do you actually run the benchmark suites in /perf?

Reply:
@gkamperis `ts-node perf/Array/copy.ts`

Reply:
@gcanti thanks! here is what I see: using the default suite I get native to be faster.
If I replace it with range(1, 3) then Array is faster...

Reply:
@gkamperis yep, in my setup, running the default suite, Array is still faster but the numbers are different (don't know why, I'm not an expert)

```
copy (native) x 65,308,620 ops/sec ±1.26% (88 runs sampled)
copy (Array) x 79,410,327 ops/sec ±0.62% (89 runs sampled)
Fastest is copy (Array)
```

For reference, `lodash` does something similar https://github.com/lodash/lodash/blob/04ebca6c86deba0ff733847f6c11fd5265e1ce03/.internal/copyArray.js

Reply:
@gcanti not an expert either...

Trying a few more combos it looks like arrays defined at compile time are treated quite differently.

So [1, ..., 49] is treated differently to [].push(1..49) 

elements: 49
copy (native) x 9,184,822 ops/sec ±3.23% (83 runs sampled)
copy (native) Lit x 121,546,767 ops/sec ±5.32% (86 runs sampled)
copy (Array) x 13,041,231 ops/sec ±1.37% (90 runs sampled)
copy (Array) Lit x 11,902,280 ops/sec ±4.37% (80 runs sampled)
Fastest is copy (native) Lit

the difference in perf is huge for arrays fully defined at compile time.